lib/segment/src/common/mmap_ops.rs

```diff
@@ -41,9 +44,105 @@ pub fn open_write_mmap(path: &Path) -> OperationResult<MmapMut> {
 
     let mmap = unsafe { MmapMut::map_mut(&file)? };
     madvise::madvise(&mmap, madvise::get_global())?;
+
     Ok(mmap)
 }
 
+#[derive(Debug)]
+pub struct PreheatDiskCacheWorker {
+    receiver: mpsc::Receiver<PreheatDiskCache>,
+}
+
+impl PreheatDiskCacheWorker {
+    pub fn spawn() -> PreheatDiskCacheHandle {
+        let (sender, receiver) = mpsc::sync_channel(32);
+
+        let handle = PreheatDiskCacheHandle { sender };
+
+        let worker = Self { receiver };
+        let _ = thread::spawn(move || worker.exec());
+
+        handle
+    }
+
+    fn exec(self) {
+        while let Ok(task) = self.receiver.recv() {
+            task.exec();
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct PreheatDiskCacheHandle {
+    sender: mpsc::SyncSender<PreheatDiskCache>,
+}
+
+impl PreheatDiskCacheHandle {
+    pub fn schedule(&self, task: PreheatDiskCache) -> OperationResult<()> {
+        self.sender.send(task).map_err(|err| {
+            OperationError::service_error(format!(
+                "failed to send preheat mmap disk-cache task{} to the worker: {err}",
+                err.0
+                    .path
+                    .as_ref()
+                    .map_or(String::new(), |path| format!(" ({path:?})")),
+            ))
+        })
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct PreheatDiskCache {
+    mmap: Arc<Mmap>,
+    path: Option<PathBuf>,
+}
+
+impl PreheatDiskCache {
+    pub fn new(mmap: Arc<Mmap>, path: Option<impl Into<PathBuf>>) -> Self {
+        Self {
+            mmap,
+            path: path.map(Into::into),
+        }
+    }
+
+    pub fn exec(&self) {
+        preheat_disk_cache(self.mmap.as_ref(), self.path.as_deref());
+    }
+}
+
+fn preheat_disk_cache<T>(mmap: &T, path: Option<&Path>)
+where
+    T: Madviseable + ops::Deref<Target = [u8]>,
+{
+    let separator = path.map_or("", |_| " "); // space if `path` is `Some` or nothing
+    let path = path.unwrap_or(Path::new("")); // path if `path` is `Some` or nothing
+
+    log::debug!("Reading mmap{separator}{path:?} to populate cache...");
+
+    let instant = time::Instant::now();
+
+    let mut dst = [0; 8096];
+
+    for iter in 0..(mmap.len() / dst.len()) {
```

Comment:
Ooooh, that's a solid advice! I always forget about `black_box`. I think we definitely want to use it here. Maybe even retry to "sparsely" read a single byte every byte through a `black_box` and see if it would work better.

Reply:
`black_box` has been unstable for a very long time, and has only recently been stabilized :smile: .

I'm not 100% this will give us our desired result though, but worth a try. If it doesn't work it may be worth taking a look at what something like [`constant_time_eq`](https://docs.rs/constant_time_eq/latest/constant_time_eq/fn.constant_time_eq.html) does internally.

> chunks also looks good. (There are too many iterator methods for me to know of. see_no_evil)

The Rust stdlib is surprisingly powerful!